tokio/src/sync/set_once.rs

```diff
@@ -0,0 +1,392 @@
+        }
+    }
+
+    /// Returns `true` if the `SetOnce` currently contains a value, and `false`
+    /// otherwise.
+    pub fn initialized(&self) -> bool {
+        // Using acquire ordering so we're able to read/catch any writes that
+        // are done with `Ordering::Release`
+        self.value_set.load(Ordering::Acquire)
+    }
+
+    // SAFETY: The SetOnce must not be empty.
+    unsafe fn get_unchecked(&self) -> &T {
+        &*self.value.with(|ptr| (*ptr).as_ptr())
+    }
+
+    /// Returns a reference to the value currently stored in the `SetOnce`, or
+    /// `None` if the `SetOnce` is empty.
+    pub fn get(&self) -> Option<&T> {
+        if self.initialized() {
+            Some(unsafe { self.get_unchecked() })
+        } else {
+            None
+        }
+    }
+
+    // SAFETY: The caller of this function needs to ensure that this function is
+    // called only when the value_set AtomicBool is flipped from FALSE to TRUE
+    // meaning that the value is being set from uinitialized to initialized via
+    // this function
+    //
+    // The caller also has to ensure writes on `value` are syncronized with a
+    // external lock to prevent mutliple set_value calls at the same time.
+    unsafe fn set_value(&self, value: T) {
+        unsafe {
+            self.value.with_mut(|ptr| (*ptr).as_mut_ptr().write(value));
+        }
+
+        // notify the waiting wakers that the value is set
+        self.notify.notify_waiters();
+    }
+
+    /// Sets the value of the `SetOnce` to the given value if the `SetOnce` is
+    /// empty.
+    ///
+    /// If the `SetOnce` already has a value, this call will fail with an
+    /// [`SetError::AlreadyInitializedError`].
+    ///
+    /// If the `SetOnce` is empty, but some other task is currently trying to
+    /// set the value, this call will fail with [`SetError::InitializingError`].
+    ///
+    /// [`SetError::AlreadyInitializedError`]: crate::sync::SetError::AlreadyInitializedError
+    /// [`SetError::InitializingError`]: crate::sync::SetError::InitializingError
+    pub fn set(&self, value: T) -> Result<(), SetError<T>> {
+        if self.initialized() {
+            return Err(SetError::AlreadyInitializedError(value));
+        }
+
+        // SAFETY: lock the mutex to ensure only one caller of set
+        // can run at a time.
+        match self.lock.lock() {
+            Ok(_) => {
+                // Using release ordering so any threads that read a true from this
+                // atomic is able to read the value we just stored.
+                if !self.value_set.swap(true, Ordering::Release) {
+                    // SAFETY: We are swapping the value_set AtomicBool from FALSE to
+                    // TRUE with it being previously false and followed by that we are
+                    // initializing the unsafe Cell field with the value
+                    unsafe {
+                        self.set_value(value);
+                    }
```

Comment:
Please set the value first, and then write to the atomic. Also, using `store` is sufficient. With the mutex, we no longer need `swap`.